cli/gitleaks/gitleaks.py

```diff
@@ -0,0 +1,312 @@
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+import logging
+import os
+import shutil
+import subprocess as sp
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any
+from typing import Literal
+
+import requests
+from simple_logger.logger import get_logger
+
+BUILD_ID_VAR: str = "BUILD_ID"
+ARTIFACT_DIR_VAR: str = "ARTIFACT_DIR"
+
+logger = get_logger(name="gitleaks")
+
+
+@dataclass
+class GitleaksConfig:
+    _server_url: str
+    _token_path: str
+    _output_file: str
+    _gitleaks_version = "7.6.1"
+    _artifact_dir: None | Path = None
+    _job_dir: None | Path = None
+    _patterns_file_url: None | str = None
+    _keep_job_dir: None | bool = False
+    _gitleaks_report_path: None | Path = None
+    _patterns_file_path: None | Path = None
+    _env_artifact_dir: str = ""
+    _env_build_id: str = ""
+    _job_dir_root: str | Path = "/tmp"
+    _logger: Any = None
+
+    def __post_init__(self) -> None:
+        """
+        Builds the GitleaksConfig object. This class is used to generate a Gitleaks detect scan using the latest Redhat
+        detection rules.
+
+        The expected environment variables, directories, and secrets are gathered and checked for availability.
+
+        The latest Gitleaks detection patterns are fetched from the Redhat patterns distribution server and used to run
+        a `gitleaks detect` scan on the job directory created by the `firewatch report` cli command.
+
+        If any detections are found, a report is created and archived in the ARTIFACT_DIR.
+
+        The detection patterns file and the job directory are removed during teardown.
+        """
+        try:
+            self._get_env_vars()
+            self._check_env()
+            self._try_fetch_patterns_from_server()
+            self._try_start_gitleaks_detect_subprocess()
+        except OSError:
+            self.logger.warning("nothing to scan. cleaning up and exiting.")
+            raise
+        finally:
+            self._cleanup()
+
+    def _get_env_vars(self) -> None:
+        """
+        Get the values for the ARTIFACT_DIR and BUILD_ID environment variables.
+        """
+        self._env_artifact_dir = os.getenv(ARTIFACT_DIR_VAR, "")
+        self._env_build_id = os.getenv(BUILD_ID_VAR, "")
+
+    def _check_env(self) -> None:
+        """
+        Check that the required environment variables, directories, and secrets are available.
+        Logs error message to the class logger object and raises OSError if one or more requirements are unmet.
+        """
+
+        def _log_err(m: str, _logger: Any = self.logger) -> Literal[True]:
+            """
+            Log the message as an error and return True. Used to toggle the _has_env_err in the outer scope.
+
+            Args:
+                m (str): The error message to log.
```

Comment:
Please don't use single character variables. Descriptive variables are preferred for readability and maintenance.